### PR TITLE
Refactoring and Documentation

### LINKED ISSUE #103: Future-proof argument matchers

### LINKED ISSUE #124: Document usage of `as_any` with mockiato
Example usage from the myelin engine:

```rust
#[cfg_attr(any(test, feature = "use-mocks"), mockable(static_references))]
pub trait ObjectBehavior: Debug + ObjectBehaviorClone + ObjectBehaviorAsAny {
    fn step(&mut self, world_interactor: &dyn WorldInteractor) -> Option<Action>;
}

pub trait ObjectBehaviorAsAny {
    fn as_any(&self) -> &'_ dyn Any;
}

impl<T> ObjectBehaviorAsAny for T
where
    T: ObjectBehavior + 'static,
{
    default fn as_any(&self) -> &'_ dyn Any {
        self
    }
}
```

### PR DESCRIPTION
**Pull request by bash**
# Resolves

- Resolves #124
- Resolves #103

# Todo
- [x] Create issue for removing proc macro documentation workaround. (#141)
- [x] Document `unordered_vec_eq`
- [x] Document `expect_` methods

### CODE REVIEW

File: crates/mockiato-codegen/src/parse/method_inputs.rs [hunk 1]
```diff
@@ -114,22 +114,25 @@ impl MethodArg {
...
             }
-            _ => Err(Error::Diagnostic(Diagnostic::spanned(
-                span,
-                Level::Error,
-                "Only captured arguments are supported",
-            ).note("This error should never appear, because rustc already enforces these requirements"))),
+            _ => Err(
+                DiagnosticBuilder::error(span, "Only captured arguments are supported")
+                    .note(
+                        "This error should never appear, because rustc already enforces these \
```
Comment (jnferner) [inline crates/mockiato-codegen/src/parse/method_inputs.rs]: Add an issue creation link on unexpected errors.

File: crates/mockiato-codegen/src/parse/mockable_attr.rs [hunk 2]
```diff
@@ -77,3 +50,54 @@ impl MockableAttr {
...
+        STATIC_REFERENCES_ATTR_PARAM_NAME
+    );
+    DiagnosticBuilder::error(meta_item.span(), error_message)
+        .build()
+        .into()
+}
+
+fn name_specified_more_than_once_error(meta_item: &Meta) -> Error {
+    let error_message = format!(
+        "`{}` should only be specified once",
```
Comment (jnferner) [inline crates/mockiato-codegen/src/parse/mockable_attr.rs]: Why is this error message only slightly different from the one in `static_references_specified_more_than_once_error`? Can't they share the phrasing?

File: crates/mockiato-codegen/src/parse/static_attr.rs [hunk 3]
```diff
@@ -1,33 +1,32 @@
...
-                "Example usage: #[{}(static_references)]",
-                ATTR_NAME
-            )),
-        ))
+        let error_message = format!(
+            "#[{}({}) does not take any parameters",
+            ATTR_NAME, STATIC_REFERENCES_ATTR_PARAM_NAME
+        );
+        let help_message = format!(
+            "Example usage: #[{}({})]",
```
Comment (jnferner) [inline crates/mockiato-codegen/src/parse/static_attr.rs]: I'd rephrase this as "Help:". It's not really an example, but a guide on how to fix the issue.

File: crates/mockiato-codegen/src/syn_ext.rs [hunk 4]
```diff
@@ -9,3 +9,28 @@ impl PathExt for Path {
...
+
+        assert!(path.first_segment_as_ident().is_none());
+    }
+
+    #[test]
+    fn returns_first_segment_without_params() {
+        let path: Path = parse_quote!(Foo<'a>::Bar::Baz);
+        let expected_ident: Ident = parse_quote!(Foo);
+
+        assert_eq!(&expected_ident, path.first_segment_as_ident().unwrap(),);
```
Comment (jnferner) [inline crates/mockiato-codegen/src/syn_ext.rs]: What is that trailing comma doing there? Shouldn't rustfmt remove it?

File: examples/clone.rs [hunk 5]
```diff
@@ -23,7 +23,7 @@ fn main() {
     let mut greeter = GreeterMock::new();
 
     greeter
-        .expect_greet(partial_eq("Tom"))
+        .expect_greet(|a| a.partial_eq("Tom"))
```
Comment (jnferner) [inline examples/clone.rs]: I think the examples would be more inviting for a new user if the closure parameter was named `arg` instead of `a`

File: examples/downcasting.rs [hunk 6]
```diff
@@ -0,0 +1,35 @@
+// In order to downcast from a trait object to a concrete implementation, a trait needs an `as_any` method.
+// [Source](https://stackoverflow.com/a/33687996/6827589)
```
Comment (jnferner) [inline examples/downcasting.rs]: You don't need an explicit `Source` link. Simply hyperlink the text above.

File: examples/downcasting.rs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+// This is an issue, because there's no way to sensibly mock this method with mockiato.
+// The solution is to split the `as_any` method into a separate trait.
+
+#[cfg(test)]
+use mockiato::mockable;
+use std::any::Any;
+
+// `static_references` is required, because the `Any` trait requires `'static`
```
Comment (jnferner) [inline examples/downcasting.rs]:
```suggestion
// `static_references` is required, because the `Any` trait requires `'static`.
```

File: src/internal/matcher/nearly_eq.rs [hunk 8]
```diff
@@ -1,29 +1,74 @@
...
+    ///
+    /// #[cfg_attr(test, mockable)]
+    /// trait FloatFormatter {
+    ///     fn format_float(&self, value: f64) -> String;
+    /// }
+    ///
+    /// let expected_result = String::from("0.3");
+    /// let mut formatter = FloatFormatterMock::new();
+    /// formatter
+    ///     .expect_format_float(|a| a.nearly_eq(0.3))
```
Comment (jnferner) [inline src/internal/matcher/nearly_eq.rs]: Again, I feel like using `arg` is nicer for beginners than `a`. Check this globally if you agree.

File: src/internal/matcher/nearly_eq.rs [hunk 9]
```diff
@@ -43,7 +88,7 @@ where
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         write!(
             f,
-            "{:?}, accuracy = {:?}",
+            "{:?}±{:?}",
```
Comment (jnferner) [inline src/internal/matcher/nearly_eq.rs]: Nice unicode 👍

File: src/internal/matcher/nearly_eq.rs [hunk 10]
```diff
@@ -82,70 +127,94 @@ mod test {
...
         let accuracy = 0.1;
 
-        assert!(!nearly_eq_with_accuracy(first_value, accuracy).matches_argument(&second_value));
+        assert!(!factory
+            .nearly_eq_with_accuracy(first_value, accuracy)
+            .matches_argument(&second_value));
     }
 
     #[test]
     fn float_is_nearly_equal_to_different_float_low_accuracy() {
```
Comment (jnferner) [inline src/internal/matcher/nearly_eq.rs]:
```suggestion
    fn float_is_nearly_equal_to_different_float_with_low_accuracy() {
```

File: src/internal/matcher/unordered_vec_eq.rs [hunk 11]
```diff
@@ -1,21 +1,41 @@
...
 use std::fmt::{self, Debug, Display};
 
-/// Creates a new `ArgumentMatcher` that matches [`Vec`]s and [`slice`]s
-/// while disregarding the exact order of the elements.
-///
-/// [`slice`]: https://doc.rust-lang.org/std/primitive.slice.html
-pub fn unordered_vec_eq<T>(vec: Vec<T>) -> UnorderedVecArgumentMatcher<T> {
-    UnorderedVecArgumentMatcher(vec)
+impl Argument {
+    /// Creates an argument matcher that against [`Vec`]s and [`slice`]s
```
Comment (jnferner) [inline src/internal/matcher/unordered_vec_eq.rs]:
```suggestion
    /// Creates an argument matcher that matches against [`Vec`]s and [`slice`]s
```

File: src/lib.rs [hunk 12]
```diff
@@ -14,13 +109,45 @@
...
 
+#[cfg(rustdoc)]
+#[macro_export]
+/// Generates a mock struct from a trait.
+///
+/// # Parameters
+///
+/// ## `static_references`
+/// Forces values stored in argument matchers to be `'static`. This is used when the mock needs to satisfy
+/// `'static` e.g. when downcasting the mocked trait to a concrete implementation using the `Any` trait.
```
Comment (jnferner) [inline src/lib.rs]: Add a link to the downcasting example.

File: crates/mockiato-codegen/src/syn_ext.rs (hunk 4 above)
Comment (bash) [reply crates/mockiato-codegen/src/syn_ext.rs]: No rustfmt doesn't remove trailing commas here 😞. Addressed in 8645106

File: src/internal/matcher/nearly_eq.rs (hunk 9 above)
Comment (bash) [reply src/internal/matcher/nearly_eq.rs]: ❤️

File: crates/mockiato-codegen/src/parse/method_inputs.rs (hunk 1 above)
Comment (bash) [reply crates/mockiato-codegen/src/parse/method_inputs.rs]: Addressed in 0211a29

File: crates/mockiato-codegen/src/parse/mockable_attr.rs (hunk 2 above)
Comment (bash) [reply crates/mockiato-codegen/src/parse/mockable_attr.rs]: Unified in 506a773

File: crates/mockiato-codegen/src/parse/static_attr.rs (hunk 3 above)
Comment (bash) [reply crates/mockiato-codegen/src/parse/static_attr.rs]: It already gets prefixed with “Help” afaik. How about “Correct usage:”?

File: examples/clone.rs (hunk 5 above)
Comment (bash) [reply examples/clone.rs]: Addressed in f22cd76fc97ee00643a49c7cec655076229458e1

File: src/internal/matcher/nearly_eq.rs (hunk 8 above)
Comment (bash) [reply src/internal/matcher/nearly_eq.rs]: Addressed in f22cd76

File: src/lib.rs (hunk 12 above)
Comment (bash) [reply src/lib.rs]: Addressed in 98ffb2c

File: examples/downcasting.rs (hunk 6 above)
Comment (bash) [reply examples/downcasting.rs]: Addressed in 9ad0699

File: crates/mockiato-codegen/src/parse/static_attr.rs (hunk 3 above)
Comment (jnferner) [reply crates/mockiato-codegen/src/parse/static_attr.rs]: Sounds good.

Comment (bash) [reply crates/mockiato-codegen/src/parse/static_attr.rs]: Addressed in 836805f
